.clj-kondo/config.edn

```diff
@@ -122,7 +123,108 @@
   {:exclude
    [colorize.core]}
 
-  :unsorted-required-namespaces {:level :warning}
+  :metabase/ns-module-checker
+  {:level :warning
+
+   ;; Map of module name => the "core" external public-facing API namespace(s).
+   ;;
+   ;; `nil` or an empty set means this module does not (yet) have an external public-facing API namespace. Go create
+   ;; one. Otherwise this should be a set of namespace symbols. All namespaces in other modules will only be allowed
+   ;; to use namespaces from this set.
+   :api-namespaces
+   {metabase.actions               nil
+    metabase.analytics             nil
+    metabase.analyze               nil
+    metabase.api                   nil
+    metabase.async                 nil
+    metabase.automagic-dashboards  nil
+    metabase.cmd                   nil
+    metabase.core                  nil
+    metabase.db                    #{metabase.db
+                                     metabase.db.metadata-queries ; FIXME, this should probably be separate from metabase.db
+                                     metabase.db.query            ; FIXME, this is mostly util stuff like `metabase.db.query/query` that we don't even need anymore.
+                                     metabase.db.setup}           ; FIXME, these are only calling `metabase.db.setup/setup-db!` and there's a slightly different version in `metabase.db`
+    metabase.domain-entities       nil
+    metabase.driver                nil
+    metabase.email                 nil
+    metabase.embed                 nil
+    metabase.events                nil
+    metabase.formatter             nil
+    metabase.integrations          nil
+    metabase.legacy-mbql           nil
+    metabase.lib                   nil
+    metabase.metabot               nil
+    metabase.models                nil
+    metabase.native-query-analyzer nil
+    metabase.permissions           nil
+    metabase.plugins               nil
+    metabase.public-settings       nil
+    metabase.pulse                 nil
+    metabase.query-processor       nil
+    metabase.search                nil
+    metabase.server                nil
+    metabase.shared                nil
+    metabase.sync                  nil
+    metabase.task                  nil
+    metabase.transforms            nil
+    metabase.types                 #{metabase.types}
+    metabase.upload                nil
+    metabase.util                  nil}
+
+   ;; Map of module => other modules you're allowed to use there.
+   ;;
+   ;; Not including a module in this list or giving it a value of `nil` means allowed modules are not enforced for
```

Comment:
Along the same line as my previous comment I would like to suggest some alternate semantics:

1. `nil` (or missing) entries mean that this is a leaf module - it cannot have any dependencies.
2. A sentinel value like `:any-modules` opts out of having any restriction on which modules can be imported.
3. A sentinel value like `:any-namespaces` opts out of encapsulation at all, e.g. out of the api-namespace restrictions as well (since these will now have a strict default).

### Alternate Suggestion:

The prior PR ended up not tackling inter-module dependencies, and had focused on just hiding internal namespaces to limit scope, but one way I'd planned to tackle this in reverse - defining "friend" namspaces for any module we don't consider usable anywhere. 

My impression is that there will be a lot of modules that we don't plan from hiding intentionally from anyone else, as long as they hide their internals, and we can make that the default in this formulation and save a lot noise and churn their entries. Another advantage this version gives is that the code will explicitly show the blast radius of changing any given module we're trying to restrict usage of, and gives a single definition to drive shrinking it.

Another idea was to just handle this through nesting, i.e. the way we'd hide say `X` from all namespaces except `A`, `B`, and `C`, would be to put them all in a given module say `M` and then propagate the visibility of only the latter namespaces up. This is roughly how ML languages typically do it.

### Related Question:

It seems like we can easily leak unused entries for either of the visibile-to or visible-from conventions. Can you think of a mechanism to lint against these dead entries? I imagine in our early days we'll accumulate a lot of these entries.

Reply:
Good idea, I added a sentinel value `:any` that will let you use any modules, otherwise the default is now `#{}`. I don't think we need `:any-namespaces`, if you really need to opt out of using this linter you can just do `clj-kondo/ignore`, but hopefully people won't ever need to do that.

Reply:
> My impression is that there will be a lot of modules that we don't plan from hiding intentionally from anyone else

My impression is the opposite, there will probably be a few foundational modules like `metabase.util` but for the most part most other modules shouldn't be used in more than a few places. Forcing you to explicitly list the allowed modules is a forcing function I think that encourages people to keep the inter-module dependencies down

Reply:
> It seems like we can easily leak unused entries for either of the visibile-to or visible-from conventions. Can you think of a mechanism to lint against these dead entries? I imagine in our early days we'll accumulate a lot of these entries.

Kondo linting hooks can't really see outside of their current namespace, to lint this effectively I think you'd have to look at the project as a whole, we could probably do that in a different way using `tools.namespace` and building up a dependency graph but I think that's outside of the scope of what I'm trying to do for now

Reply:
> Another idea was to just handle this through nesting, i.e. the way we'd hide say `X` from all namespaces except `A`, `B`, and `C`, would be to put them all in a given module say `M` and then propagate the visibility of only the latter namespaces up. This is roughly how ML languages typically do it.

That's one option, I'm trying to make this PR agnostic about how the modules end up looking, the main goal here is just to provide us with a framework so we can actually start putting things into modules and enforcing them. This linter should leave us open to do things that way if we want, or not if we don't, shouldn't make a difference either way I think.